File: adventofcode-2024/src/main/java/nl/mvdr/adventofcode/adventofcode2024/day22/Part2.java
package nl.mvdr.adventofcode.adventofcode2024.day22;

import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import nl.mvdr.adventofcode.solver.LongSolver;

public class Part2 implements LongSolver {

    private static final Logger LOGGER = LoggerFactory.getLogger(Part2.class);

    @Override
    public long solve(Stream<String> lines) {
        var secretNumberSequences = lines.mapToLong(Long::parseLong)
                .mapToObj(SecretNumber::new)
                .map(SecretNumber::generateAll)
                .collect(Collectors.toSet());
        
        return secretNumberSequences.stream()
                .flatMap(sequence -> sequence.stream()
                        .map(SecretNumber::priceChanges))
                .filter(priceChangeSequence -> priceChangeSequence.size() == 4)
                .distinct()
                .mapToLong(priceChangeSequence -> secretNumberSequences.stream()
                        .parallel()
                        .mapToLong(secretNumberSequence -> secretNumberSequence.stream()
                                .filter(secretNumber -> secretNumber.priceChanges().equals(priceChangeSequence))
                                .mapToLong(SecretNumber::price)
                                .findFirst()
                                .orElse(0L))
                        .sum())
                .max()
                .orElseThrow();
    }
    
    public static void main(String[] args) {
        var instance = new Part2();

        var result = instance.solve("input-day22-2024.txt");

        LOGGER.info(result);
    }
}
 

File: adventofcode-2025/src/main/kotlin/nl/mvdr/adventofcode/adventofcode2025/day03/Part2.kt
package nl.mvdr.adventofcode.adventofcode2025.day03

import io.github.oshai.kotlinlogging.KotlinLogging
import nl.mvdr.adventofcode.solver.FunctionSolver

private val logger = KotlinLogging.logger{}

fun solvePart2(lines: Sequence<String>): Long =
    lines.map(::parseBank)
        .map { it.computeMaxJoltage(12) }
        .sum()

fun main() {
    val result = FunctionSolver(::solvePart2).solve("input-day03.txt")
    logger.info { result }
}
